src/tools/wasm-merge.cpp

```diff
@@ -0,0 +1,527 @@
+    auto copy = std::make_unique<Export>(*curr);
+
+    // An export may already exist with that name, so fix it up.
+    copy->name = Names::getValidExportName(merged, copy->name);
+
+    // Note the module origin and original name of this export, for later fusing
+    // of imports to exports.
+    exportModuleMap[copy.get()] = ExportInfo{inputName, curr->name};
+
+    // Add the export.
+    merged.addExport(std::move(copy));
+  }
+
+  // Start functions must be merged.
+  if (input.start.is()) {
+    if (!merged.start.is()) {
+      // No previous start; just refer to the new one.
+      merged.start = input.start;
+    } else {
+      // Merge them, keeping the order. Note that we need to create a new
+      // function as there may be other references.
+      Builder builder(merged);
+      auto mergedName = Names::getValidFunctionName(merged, "merged.start");
+      auto* oldStart = merged.getFunction(merged.start);
+      auto* oldStartBody = ExpressionManipulator::copy(oldStart->body, merged);
+      auto* newStart = merged.getFunction(input.start);
+      auto* newStartBody = ExpressionManipulator::copy(newStart->body, merged);
+      auto* mergedBody = builder.makeSequence(oldStartBody, newStartBody);
+      auto mergedFunc = builder.makeFunction(
+        mergedName, Signature{Type::none, Type::none}, {}, mergedBody);
+      merged.addFunction(std::move(mergedFunc));
+      merged.start = mergedName;
+    }
+  }
+
+  // TODO: type names, features, debug info, custom sections, dylink info, etc.
+}
+
+// Find pairs of matching imports and exports, and make uses of the import refer
+// to the exported item (which has been merged into the module).
+void fuseImportsAndExports() {
+  // First, scan the exports and build a map. We build a map of [module name] to
+  // [export name => internal name]. For example, consider this module:
+  //
+  //  (module ;; linked in as "module_A"
+  //    (func $foo (export "bar"))
+  //  )
+  //
+  // Then the ModuleExportMap will be:
+  //
+  //  {
+  //    "module_A": {
+  //      "bar": "foo";
+  //    }
+  //  }
+  //
+  using ModuleExportMap = std::unordered_map<Name, NameMap>;
+
+  // A map of ModuleExportMaps, one per item kind (one for functions, one for
+  // globals, etc.).
+  using KindModuleExportMaps =
+    std::unordered_map<ExternalKind, ModuleExportMap>;
+  KindModuleExportMaps kindModuleExportMaps;
+
+  for (auto& ex : merged.exports) {
+    assert(exportModuleMap.count(ex.get()));
+    ExportInfo& exportInfo = exportModuleMap[ex.get()];
+    kindModuleExportMaps[ex->kind][exportInfo.moduleName][exportInfo.baseName] =
+      ex->value;
+  }
+
+  // Find all the imports and see which have corresponding exports, which means
+  // there is an internal item we can refer to. We build up a map of the names
+  // that we should update.
+  KindNameMaps kindNameMaps;
+  ModuleUtils::iterImportable(merged, [&](ExternalKind kind, Importable* curr) {
+    if (curr->imported()) {
```

Comment:
We don't have an iterator utility that does this check for us?

Reply:
We could add one. We have ones for imported/defined functions, etc. But as this may be the only place that uses it, it might be excessive.

Reply:
I agree that this single use case probably doesn't warrant a new utility.